siq integration

ScottSoren opened the pull request:
This PR resolves https://github.com/ixdat/ixdat/issues/122.

I've put the prefix "siq_" in front of the MS calibration methods that use ``spectro_inlets_quantification``. I've also moved the native MS calibration methods from ``MSInlet`` to ``MSMeasurement``. I've left deprecated methods in MSInlet which call the MSMeasurement methods so that all scripts using the native ixdat calibration the old way will keep working for now.
I've also put `siq_` in front of everything to do with siq, which meant replacing `si_quant` where that had been used.

It's demo'd in development_scripts/siq_integration_demo which makes use of data in the ixdat-large-test-files submodule.

Should be a quick and easy review since we discussed the issue beforehand.
Once it's merged, however, I will (as promised) make a new issue calling for the spinning out of molecule data to a separate repo.

AlexanderKrabbe commented on src/ixdat/techniques/ms.py:
```diff
@@ -414,8 +449,11 @@ def gas_flux_calibration_curve(
-                ms_calibration on. Defaults to None. These axes are not returned.
+            axis_measurement (Axis): The MS plot axes to highlight the
+                ms_calibration on. Defaults to None.
+            tspans for ms_calibration is showing raw data or bg subtracted data
```
I dont think this "tspans" is a defined argument in the function. So it can not be changed from the default "False".

AnnaWiniwarter replied:
Seems like you missed or deleted some lines when copying the method.

ScottSoren replied:
Addressed, answer in response to Anna's other comment.

AlexanderKrabbe commented on src/ixdat/techniques/ms.py:
```diff
@@ -424,18 +462,8 @@ def gas_flux_calibration_curve(
-        CalPoint = plugins.si_quant.CalPoint
-
-        chip = chip or Chip()
+        # prepare three lists to loop over to determine molecule flux in the
```
Line 465 and 466 looks identical to 468 and 469

ScottSoren replied:
Yes. fixed.

AlexanderKrabbe commented on src/ixdat/techniques/ms.py:
```diff
@@ -348,41 +348,74 @@ def as_mass(self, item):
 
-    def gas_flux_calibration(self, mol, mass, tspan, chip=None):
-        """Simple pure-gas flux calibration, using external MS quantification package
+    def gas_flux_calibration(
```
Would it make sense to add a "warning" if one has activated a plugin but mistakenly useds this function instead of sig_gas_flux_calibration? Or should this function simply just pipe through to external quantification packages in case those af active?

AnnaWiniwarter replied:
Kenneth, Soren and me had a long discussion on why we don't want it to just "pipe through" to the external package (in a nutshell: because it makes it very hard to understand what's going on if you dont really know what you're doing). I think it's a good idea with a warning though, especially since at the moment it just goes to the external package and this change will be a bit confusing for existing users

ScottSoren replied:
I've put a warning here. It would make sense to put it in `gas_flux_calibration_curve` as well but I can't as it is now since `siq_gas_flux_calibration_curve` calls `gas_flux_calibration_curve`. Not sure if that inconsistency just makes it worse?

AnnaWiniwarter replied:
Can `gas_flux_calibration_curve` figure out that it's being called from `siq_gas_flux_calibration_curve` and disable a warning in that case? Or does it become too complex then?
I dont think that the inconsistency, but we need to make sure that the documentation is good enough.

AlexanderKrabbe replied:
Or another way would be to place the code of `gas_flux_calibration_curve()` in and internal function of the class  `_gas_flux_calibration_curve()`. 

```
gas_flux_calibration_curve():
     if external_package_active:
        warn.warning('You have activated an external package. Call this function instead 'siq_gas_flux_calibration_curve')

    _gas_flux_calibration()
```

AlexanderKrabbe commented on src/ixdat/techniques/ms.py:
```diff
@@ -500,12 +528,126 @@ def gas_flux_calibration_curve(
+        return cal
+
+    def siq_gas_flux_calibration(self, mol, mass, tspan, chip=None):
+        """Simple pure-gas flux calibration, using external MS quantification package
```
it is a highly specialised external package right? Or would this function be used with any external MS quantification pacakge

ScottSoren replied:
Good point. "external MS quantification package" replaced with "`spectro_inlets_quantification`" here and elsewhere.

AlexanderKrabbe commented on src/ixdat/techniques/ms.py:
```diff
@@ -348,41 +348,74 @@ def as_mass(self, item):
+        mol_conc_ppm=None,
+    ):
+        """
+        Fit mol's sensitivity at mass based on period with steady gas composition.
```
Is is "based on one period" correct?

AlexanderKrabbe reviewed (commented):
@ScottSoren Some comments are grammar, some are programming. I need my personal computer to test the changes in work flow on true data. 

One thing I think could be nice is to discuss is how well integrated sig as an external quantification package actually is and how this would be in the future if other external quantification packages arises. How much should IXDAT plugin from those packages and how much should just be placed in the external package. 

It is the function `set_sig_quantifier() `  and `_with_siq_quantifier` [here](https://github.com/ixdat/ixdat/pull/137/commits/367926194d053f4c1bf1bde434742d7c9d1035c4#diff-a18a9263f34d117472ca29ea98e800784283d50630674a9512bf45a8075ee3bcR754) and [here](https://github.com/ixdat/ixdat/pull/137/commits/367926194d053f4c1bf1bde434742d7c9d1035c4#diff-a18a9263f34d117472ca29ea98e800784283d50630674a9512bf45a8075ee3bcR27)

AnnaWiniwarter commented on src/ixdat/techniques/ms.py:
```diff
@@ -500,12 +528,126 @@ def gas_flux_calibration_curve(
+                "`MSMeasurement.gas_flux_calibration_siq` only works when using an "
+                "external MS quantification package "
+                "(`ixdat.options.activate_siq()`). "
+                "For native ixdat MS quantification, `gas_flux_calibration` has to be"
```
This is outdated (for native ixdat MS quantification, the function has to be called without the siq)

ScottSoren replied:
fixed.

AnnaWiniwarter commented on src/ixdat/techniques/ms.py:
```diff
@@ -500,12 +528,126 @@ def gas_flux_calibration_curve(
+            return_ax (bool): Whether to return the axis on which the calibration is
+                plotted together with the MSCalResult. Defaults to False.
+
+        Return MSCalResult(, Axis): The result of the MS calibration (and calibration
```
This returns an `CalPoint` object!!

AnnaWiniwarter commented on src/ixdat/techniques/ms.py:
```diff
@@ -500,12 +528,126 @@ def gas_flux_calibration_curve(
+                "`MSMeasurement.gas_flux_calibration` only works when using an "
+                "external MS quantification package "
+                "(`ixdat.options.activate_siq()`). "
+                "For native ixdat MS quantification, `gas_flux_calibration` has to be"
```
outdated error message, see above

ScottSoren replied:
fixed.

AnnaWiniwarter commented on src/ixdat/techniques/ms.py:
```diff
@@ -1123,122 +1230,22 @@ def gas_flux_calibration_curve(
         tspan_bg=None,
         ax="new",
-        axes_measurement=None,
-        axes_measurement_raw=False,
```
you dropped that in the new method, but I'm not sure why? I just added this option in one of the more recent PRs to solve issues. We agreed that in the future this should be done better, but I can't see that you added a better option. Did I miss something there?

ScottSoren replied:
Got it. Sorry. Shouldn't have dropped. I couldn't remember or quickly tell what it was. Reasons for that include:
- The indentation in the docstring was off so I couldn't immediately tell how much was part of the description for this variable.
- The name was wierd since it is a bool and not a list of axes. I've now given it a name `remove_bg_on_axis_measurement` which I think better representes the type and what it does.
- it was in the native but not siq version of the function
- ax**e**s_measurement was confusing since it is a single axis

I think I've fixed all of this now, and the option is there again! :)

AnnaWiniwarter reviewed (changes requested):
I think this PR will make quantification and the differentiation between using native ixdat functions and the siq plugin much clearer. And I also appreciate very much that the calibration methods are now methods of the measurement rather than the inlet. I think usability a lot more straightforward like this. 
There are a few places where you missed some things when copying over code, but all changes should be easy to implement. 

Tested a basic single point calibration using both methods (native ixdat and siq) and it seems to work fine. Looking forward to having this merged. :)

One comment on error messages though: When accidentally running `meas.gas_flux_calibration()` and passing a `chip` (as would be the syntax for the siq-based method) the error is quite unspecifically pointing out that it doesnt know what a chip is. any chance we can add a more specific error message to the method here at least for now that it's new?

ScottSoren commented:
Thanks for the reviews, @AnnaWiniwarter  and @AlexanderKrabbe !

I've addressed all comments. Only this one was left un-resolved: https://github.com/ixdat/ixdat/pull/137#discussion_r1388002984

Otherwise, let me know if it looks better!

Also, I want to add methods for converting between the ixdat and siq calibration objects, i.e.
ixdat.techniques.ms.MSCalResult <--> spectro_inlets_quantification.calibration.CalPoint
and 
ixdat.techniques.ms.MSCalibration <--> spectro_inlets_quantification.calibration.Calibration

Any thoughts on how you'd like the API of those methods to be? How might you want to call them and where might you want to import them from?

AnnaWiniwarter reviewed (changes requested):
Thanks for putting the option to not BG subtract the current in the highlights for the calibration curve back in. Sorry about the confusing name and missing tabs in the docstring. 

I'm still not quite satisfied with the error messages, but getting there ;) this will be a great improvement I think.

AnnaWiniwarter commented:
Regarding 

> Also, I want to add methods for converting between the ixdat and siq calibration objects, i.e.
ixdat.techniques.ms.MSCalResult <--> spectro_inlets_quantification.calibration.CalPoint
and
ixdat.techniques.ms.MSCalibration <--> spectro_inlets_quantification.calibration.Calibration

> Any thoughts on how you'd like the API of those methods to be? How might you want to call them and where might you want to import them from?

So far I haven't really saved and imported calibrations (because I've been too lazy to look up the API), but I suppose I'd save them with the raw data (and thus also load them from the folder where I have the calibration data).
Regarding API, I don't know, something simple and self-explanatory? my_calibration.convert_to_MSCalResult or  my_calibration.convert_to_CalPoint?

ScottSoren commented on src/ixdat/techniques/ms.py:
```diff
@@ -241,7 +241,7 @@ def grab_siq_fluxes(
+                "`spectro_inlets_quantification` "
                 "(`ixdat.plugins.activate_siq()`). "
                 "For native ixdat MS quantification, `gas_flux_calibration` has to be"
                 "called from an instance of `MSInlet`."
```
This one slipped through. Fixed now.

ScottSoren commented:
This PR also fixes a tutorial which I had sloppily updated according to the changes on this branch and added with crashed output to the documentation page:
- Main: https://ixdat.readthedocs.io/en/latest/tutorials/tutorials_repo/L2_techniques/ec_ms_quantification/ixdat_demo_EC-MS.html#Gas-calibration
- This branch: https://ixdat.readthedocs.io/en/siq_integration/tutorials/tutorials_repo/L2_techniques/ec_ms_quantification/ixdat_demo_EC-MS.html#Gas-calibration

ScottSoren commented:
Hi @AnnaWiniwarter and @AlexanderKrabbe , 

This is ready for your review again, hopefully ready now to merge!
I've added the interconversion functions between siq and ixdat Calibration objects. They are demo'd here: 

https://github.com/ixdat/ixdat/blob/6964e63ce52d8e527ffc7a97e8b1cb72f6e0f882/development_scripts/demo_ecms_calibration_curve.py#L68

NOTE: As mentioned in my above comment, this fixes a tutorial in the docs, thus the big diff on an ipynb. This due to overlooking when merging the documentation PR that the tutorial had assumed that the siq integration PR was already merged. There was no change in the un-compiled .ipynb only the output. It was an unusal case and *not* normal workflow, which is to develop tutorials without output in the tutorials repo, not in ixdat/docs.

AnnaWiniwarter reviewed (approved):
From what I can see at a quickl glance this looks good to me now! Thanks for implementing all my suggestions @ScottSoren!

AlexanderKrabbe commented on development_scripts/demo_ecms_calibration_curve.py:
```diff
@@ -48,16 +48,41 @@
+print(siq_calibration)
+
+# The following dowsn't work because it's a SensitivityList instead of a Calibration :(
+# siq_calibratoin.plot_as_spectrum()
```
# siq_calibration.plot_as_spectrum() (spelling)

ScottSoren replied:
Thanks for the typo catch! 
Yeah, of course, one would never actually do this. It was just to show that all of the interconversions work!
